testsuite/tests/statmemprof/arrays_in_major.ml

```diff
@@ -0,0 +1,142 @@
+(* TEST
+   flags = "-g"
+   * bytecode
+     reference = "${test_source_directory}/arrays_in_major.byte.reference"
+   * native
+     reference = "${test_source_directory}/arrays_in_major.opt.reference"
+     compare_programs = "false"
+*)
+
+open Gc.Memprof
+
+let root = ref []
+let[@inline never] allocate_arrays lo hi cnt keep =
+  for j = 0 to cnt-1 do
+    for i = lo to hi do
+      root := Array.make i 0 :: !root
+    done;
+    if not keep then root := []
+  done
+
+let check_nosample () =
+  Printf.printf "check_nosample\n%!";
+  start {
+      sampling_rate = 0.;
+      callstack_size = 10;
+      callback = fun _ ->
+        Printf.printf "Callback called with sampling_rate = 0\n";
+        assert(false)
+  };
+  allocate_arrays 300 3000 1 false
+
+let () = check_nosample ()
+
+let check_ephe_full () =
+  Printf.printf "check_ephe_full\n%!";
+  let ephes = ref [] in
+  start {
+    sampling_rate = 0.01;
+    callstack_size = 10;
+    callback = fun _ ->
+      let res = Ephemeron.K1.create () in
+      ephes := res :: !ephes;
+      Some res
+  };
+  allocate_arrays 300 3000 1 true;
+  stop ();
+  List.iter (fun e -> assert (Ephemeron.K1.check_key e)) !ephes;
+  Gc.full_major ();
+  List.iter (fun e -> assert (Ephemeron.K1.check_key e)) !ephes;
+  root := [];
+  Gc.full_major ();
+  List.iter (fun e -> assert (not (Ephemeron.K1.check_key e))) !ephes
+
+let () = check_ephe_full ()
+
+let check_no_nested () =
+  Printf.printf "check_no_nested\n%!";
+  let in_callback = ref false in
+  start {
+      sampling_rate = 1.;
+      callstack_size = 10;
+      callback = fun _ ->
+        assert (not !in_callback);
+        in_callback := true;
+        allocate_arrays 300 300 100 false;
+        in_callback := false;
+        None
+  };
+  allocate_arrays 300 300 100 false;
+  stop ()
+
+let () = check_no_nested ()
+
+let check_distrib lo hi cnt rate =
+  Printf.printf "check_distrib %d %d %d %f\n%!" lo hi cnt rate;
+  let smp = ref 0 in
+  start {
+      sampling_rate = rate;
+      callstack_size = 10;
+      callback = fun info ->
+        assert (info.kind = Major);
+        assert (info.tag = 0);
+        assert (info.size >= lo && info.size <= hi);
+        assert (info.n_samples > 0);
+        smp := !smp + info.n_samples;
+        None
+    };
+  allocate_arrays lo hi cnt false;
+  stop ();
+
+  (* The probability distribution of the number of samples follows a
+     poisson distribution with mean tot_alloc*rate. Given that we
+     expect this quantity to be large (i.e., > 100), this distribution
+     is approximately equal to a normal distribution. We compute a
+     1e-6 confidence interval for !smp using quantiles of the normal
+     distribution, and check that we are in this confidence interval. *)
+  let tot_alloc = cnt*(lo+hi+2)*(hi-lo+1)/2 in
+  let mean = float tot_alloc *. rate in
+  let stddev = sqrt mean in
+  (* This assertion has probability to fail close to 1e-6. *)
```

Comment:
A `1e-6` failure probability is a bit high for a check that is run hundreds of times a day on our CI machines (each test run performs this check 6 times, we have maybe a dozen PR pushes a day that run the CI, and about 5 testsuite runs per CI run; if I count right, that's about 130K checks a year). Would it hurt the reliability of the check (to detest broken distributions) to move to 1e-10 or 1e-12?

Reply:
To be fair, this test is already rather tolerant: You have to remember that the sampling mechanism is deterministic, in the sense that two consecutive executions of the same program will end up in the exact same samples. Essentially, we will get different samples only when someone tests changes that affect the way allocation is performed *during the test*.

Reply:
It would hurt reliability a bit since it requires to increase the size of the confidence interval from 4.9 to 6.5 standard deviations (for 1e-10) (if the gaussian approximation is still accurate at this precision). In particular, this implies that the first test

```OCaml
 check_distrib 300 3000 1 0.00001;
```
would lost most of its (low) discriminating power since it would test that the empirical mean belongs to `[2,88]`. But the other test which have a far higher mean  would be less affected.

Reply:
To cut the pear in half and reassure @gasche, I've changed the probability to about 10^-8, by using a coefficient of 5.7. The test @Octachron is speaking about still has a bit of discriminating power.

Reply:
I was already fairly reassured by your point on deterministic sampling, but let's say I feel very very safe now. That said, I don't regret taking some time to consider the issue, because non-deterministic test failures are horrible.